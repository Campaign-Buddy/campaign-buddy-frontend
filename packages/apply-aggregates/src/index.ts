export { applyAggregates } from './applyAggregates';
export { getFullAggregates } from './getFullAggregates';
