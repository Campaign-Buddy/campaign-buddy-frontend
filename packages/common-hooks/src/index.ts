export * from './useBooleanState';
export * from './useCancelableCallback';
export * from './useDebouncedCallback';
export * from './useDebouncedEffect';
export * from './useIsMounted';
export * from './useStableValue';
export * from './useUpdatingRef';
export * from './useResizeObserver';
export * from './useCombinedRefs';
