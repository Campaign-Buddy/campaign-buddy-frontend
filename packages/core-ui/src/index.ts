export { Input, TextArea, NumberInput } from './input';
export { Switch } from './switch';
export { Button } from './button';
export { Modal } from './modal';
