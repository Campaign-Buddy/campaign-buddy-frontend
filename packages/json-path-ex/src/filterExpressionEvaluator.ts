/**
 * Copied from https://blog.risingstack.com/writing-a-javascript-framework-sandboxed-code-evaluation/
 */
function compileCode(src: string) {
	src = `with (sandbox) {${src}}`;
	const code = new Function('sandbox', src);

	return function (sandbox: any) {
		const sandboxProxy = new Proxy(sandbox, { has, get });
		return code(sandboxProxy);
	};
}

function has() {
	return true;
}

function get(target: any, key: any) {
	if (key === Symbol.unscopables) return undefined;
	return target[key];
}

export function evaluateFilterExpression(
	expression: string,
	currentElement: any,
	root: any
): boolean {
	if (expression.includes('return')) {
		throw new Error('Filter expression must not contain return statement');
	}

	if (/[^\w\d]_\$[^\w\d]|^_\$[^\w\d]|[^\w\d]_\$$|^_\$$/.test(expression)) {
		throw new Error(
			"Filter expression must not contain '_$'. It is reserved for the internal filter expression implementation."
		);
	}

	const normalizedExpression = `return ${expression.replace(/@/g, '_$')}`;

	try {
		return Boolean(
			compileCode(normalizedExpression)({ ['_$']: currentElement, ['$']: root })
		);
	} catch {
		return false;
	}
}
