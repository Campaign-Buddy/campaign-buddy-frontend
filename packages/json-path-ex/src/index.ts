export * from './query';
export * from './analyzeQuery';
export { QueryExpressionKind } from './syntaxAnalyzer';
