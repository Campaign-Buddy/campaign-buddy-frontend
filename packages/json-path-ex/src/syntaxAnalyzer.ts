export enum QueryExpressionKind {
  RecursiveDescent,
  ChildDescent,
  Root,
  ValueFilter,
  KeyFilter,
  PropertyAccessor,
  WildCard,
  Slice,
}

interface QueryExpression {
  kind: QueryExpressionKind;
  content: string;
}

export function popQueryExpression(
	query: string
): [expression: QueryExpression, remaining: string] | undefined {
	if (!query) {
		return;
	}

	if (query.startsWith('..')) {
		return [
			{
				kind: QueryExpressionKind.RecursiveDescent,
				content: '..',
			},
			query.substring(2),
		];
	}

	if (query.startsWith('.')) {
		return [
			{
				kind: QueryExpressionKind.ChildDescent,
				content: '.',
			},
			query.substring(1),
		];
	}

	if (query.startsWith('$')) {
		return [
			{
				kind: QueryExpressionKind.Root,
				content: '$',
			},
			query.substring(1),
		];
	}

	if (query.startsWith('*')) {
		return [
			{
				kind: QueryExpressionKind.WildCard,
				content: '*',
			},
			query.substring(1),
		];
	}

	if (query.startsWith('[?(')) {
		const match = /^\[\?\(([^)]+)\)\]/.exec(query);

		if (!match) {
			throw new Error('Invalid value filter expression');
		}

		return [
			{
				kind: QueryExpressionKind.ValueFilter,
				content: match[1],
			},
			query.substring(match[0].length),
		];
	}

	if (query.startsWith('[')) {
		const sliceMatch = /^\[(\d+)?:(\d+)?(?::(\d+))?\]/.exec(query);

		if (sliceMatch) {
			const [match, start, end, increment] = sliceMatch;

			return [
				{
					kind: QueryExpressionKind.Slice,
					content: `${start ?? '0'}:${end ?? ''}:${increment ?? '1'}`,
				},
				query.substring(match.length),
			];
		}

		const propertyAccessor = /^\[((?:'|")?)([^[\]'"]+)\1\]/.exec(query);

		if (!propertyAccessor) {
			throw new Error('Invalid property accessor expression');
		}

		return [
			{
				kind: QueryExpressionKind.PropertyAccessor,
				content: propertyAccessor[2],
			},
			query.substring(propertyAccessor[0].length),
		];
	}

	if (query.startsWith('<?(')) {
		const match = /^<\?\(([^)]+)\)>/.exec(query);

		if (!match) {
			throw new Error('Invalid key filter expression');
		}

		return [
			{
				kind: QueryExpressionKind.KeyFilter,
				content: match[1],
			},
			query.substring(match[0].length),
		];
	}

	const match = /^[^{}.:@<>?()@$[\]]+/.exec(query);

	if (!match) {
		throw new Error('Invalid property accessor');
	}

	if (/\s/.test(match[0])) {
		throw new Error('Direct property accessor must not contain whitespace');
	}

	return [
		{
			kind: QueryExpressionKind.PropertyAccessor,
			content: match[0],
		},
		query.substring(match[0].length),
	];
}
