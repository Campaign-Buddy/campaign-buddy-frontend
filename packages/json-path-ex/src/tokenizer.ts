enum TokenKind {
	// Keyword tokens
	Root,
	ChildSelector,
	RecursiveDescent,
	CurrentElement,
	Colon,

	// Grouping Tokens
	BracketStart,
	BracketEnd,
	AngleBracketStart,
	AngleBracketEnd,
	FilterExpressionStart,
	ExpressionStart,
	ExpressionEnd,

	// Unrecognized tokens,
	// could be an accessor
	ExpressionContent,
}

function getNextToken(query: string): [kind: TokenKind, content: string, remainingQuery: string] | undefined {
	if (!query) {
		return;
	}

	if (query.startsWith('$')) {
		return [TokenKind.Root, '$', query.substring(1)];
	}

	if (query.startsWith('..')) {
		return [TokenKind.RecursiveDescent, '..', query.substring(2)];
	}

	if (query.startsWith('.')) {
		return [TokenKind.ChildSelector, '.', query.substring(1)];
	}

	if (query.startsWith('@')) {
		return [TokenKind.CurrentElement, '@', query.substring(1)];
	}

	if (query.startsWith(':')) {
		return [TokenKind.Colon, ':', query.substring(1)];
	}

	if (query.startsWith('[')) {
		return [TokenKind.BracketStart, '[', query.substring(1)];
	}

	if (query.startsWith(']')) {
		return [TokenKind.BracketEnd, ']', query.substring(1)];
	}

	if (query.startsWith('<')) {
		return [TokenKind.AngleBracketStart, '<', query.substring(1)];
	}

	if (query.startsWith('>')) {
		return [TokenKind.AngleBracketEnd, '>', query.substring(1)];
	}

	if (query.startsWith('?(')) {
		return [TokenKind.FilterExpressionStart, '?(', query.substring(2)];
	}

	if (query.startsWith('(')) {
		return [TokenKind.ExpressionStart, '(', query.substring(1)];
	}

	if (query.startsWith(')')) {
		return [TokenKind.ExpressionEnd, ')', query.substring(1)];
	}

	const match = /^[^\.:@\<\>\?\(\)@$\[\]]+/.exec(query);

	if (!match) {
		throw new Error(`Unexpected token: ${query}`);
	}

	return [TokenKind.ExpressionContent, match[0], query.substring(match[0].length)];
}

export function tokenize(query: string): [kind: TokenKind, content: string][] {
	if (!query) {
		return [];
	}

	const allResults: [TokenKind, string][] = [];
	let cur = getNextToken(query);
	while (cur) {
		const [kind, content, remaining] = cur;
		allResults.push([kind, content]);
		cur = getNextToken(remaining);
	}

	return allResults;
}
