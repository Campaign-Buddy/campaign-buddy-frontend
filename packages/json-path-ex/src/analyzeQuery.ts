import { resolveSubQueries } from './query';
import { QueryExpressionKind } from './syntaxAnalyzer';

interface AnalysisResult {
	evaluatedExpressions: {
		path: string;
		expressionKind: QueryExpressionKind;
	}[];
}

export function analyzeQuery(json: any, q: string): AnalysisResult {
	return analyzeSubQueries(json, `{${q}}`);
}

export function analyzeSubQueries(json: any, q: string): AnalysisResult {
	const result: AnalysisResult = {
		evaluatedExpressions: [],
	};

	resolveSubQueries(json, q, {
		onEvaluatingExpression: (paths, expressionKind) => {
			result.evaluatedExpressions.push(
				...paths.map((path) => ({
					path,
					expressionKind,
				}))
			);
		},
	});

	return result;
}
