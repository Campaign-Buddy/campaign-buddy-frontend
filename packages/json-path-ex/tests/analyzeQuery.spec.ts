import { analyzeQuery } from '../src';

interface TestCase {
	query: string;
	json: any;
}

const testCases: TestCase[] = [
	{
		query: '$.foo',
		json: {
			foo: 'f',
			baz: 'b',
		},
	},
	{
		query: '$.foo[?(@ > 2)]',
		json: {
			foo: [0, 2, 3, 4],
			baz: 'b',
		},
	},
	{
		query: '$.foo<?(@ === "bar")>',
		json: {
			foo: {
				bar: 'baar',
				bop: 'boop',
			},
			baz: 'b',
		},
	},
	{
		query: '$.foo[?(@.id > 0)]..bings',
		json: {
			foo: [
				{ id: 0, nested: { bings: 'ruhroh' } },
				{ id: 1, nested: { bings: 'okay' } },
				{ id: 2, nested: { bings: 'okay' } },
				{ id: 3, nested: { bings: 'okay' } },
			],
		},
	},
	{
		query: '$.foo..bings',
		json: {
			foo: [
				{ id: 0, nested: { bings: 'ruhroh' } },
				{ id: 1, nested: { bings: 'okay' } },
				{ id: 2, nested: { bings: 'okay' } },
				{ id: 3, nested: { bings: 'okay' } },
			],
		},
	},
];

describe('analyzeQuery', () => {
	for (const testCase of testCases) {
		it(`query analysis works for query = ${testCase.query}`, () => {
			const result = analyzeQuery(testCase.json, testCase.query);
			expect(result).toMatchSnapshot();
		});
	}
});
