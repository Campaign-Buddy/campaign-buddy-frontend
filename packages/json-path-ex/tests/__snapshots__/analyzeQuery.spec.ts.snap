// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analyzeQuery query analysis works for query = $.foo 1`] = `
Object {
  "evaluatedExpressions": Array [],
}
`;

exports[`analyzeQuery query analysis works for query = $.foo..bings 1`] = `
Object {
  "evaluatedExpressions": Array [],
}
`;

exports[`analyzeQuery query analysis works for query = $.foo<?(@ === "bar")> 1`] = `
Object {
  "evaluatedExpressions": Array [],
}
`;

exports[`analyzeQuery query analysis works for query = $.foo[?(@ > 2)] 1`] = `
Object {
  "evaluatedExpressions": Array [],
}
`;

exports[`analyzeQuery query analysis works for query = $.foo[?(@.id > 0)]..bings 1`] = `
Object {
  "evaluatedExpressions": Array [],
}
`;

exports[`analyzeQuery query analysis works for query = {$.foo.bar} + {$.baz.bop} 1`] = `
Object {
  "evaluatedExpressions": Array [
    Object {
      "expressionKind": "root",
      "path": "$",
    },
    Object {
      "expressionKind": "childDescent",
      "path": "$",
    },
    Object {
      "expressionKind": "propertyAccessor",
      "path": "$",
    },
    Object {
      "expressionKind": "childDescent",
      "path": "$.foo",
    },
    Object {
      "expressionKind": "propertyAccessor",
      "path": "$.foo",
    },
  ],
}
`;
